### PR TITLE
Add method exists condition for testing

### PR DESCRIPTION
**Pull request by Pikachu920**
### Description
Adds a method exists condition to be used with test cases. This can be used as an alternative to the typical minecraft version checking pattern tests currently use. Examples can be seen in the test script (also below):
```
test "method exists":
	assert method "org.bukkit.Bukkit##getServer()" exists with "Existing method was not found"
	assert method "org.fake.class##notReal()" doesn't exist with "Fake method was found"
	assert method "org.bukkit.Bukkit##getEntity(java.util.UUID)" exists with "Existing method with args was not found"
```

---
**Target Minecraft Versions:** Any
**Requirements:**: Any
**Related Issues:** N/A

### CODE REVIEW

File: src/main/java/ch/njol/skript/tests/runner/CondMethodExists.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+@Name("Method Exists")
+@Description("Checks if a method exists")
+@Examples("if method \"org.bukkit.Bukkit#getPluginCommand(java.lang.String)")
+@Since("INSERT VERSION")
+public class CondMethodExists extends PropertyCondition<String> {
+
+	private final static Pattern SIGNATURE_PATTERN = Pattern.compile("(?<class>.+)#(?<name>.+)\\((?<params>.*)\\)");
+
+	static {
+		Skript.registerCondition(CondMethodExists.class, "method[s] %strings% [(1¦don't|1¦doesn't)] exist[s]");
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/tests/runner/CondMethodExists.java:48]:
```suggestion
		Skript.registerCondition(CondMethodExists.class, "method[s] %strings% [dont:do(esn't|n't)] exist[s]");
```
and `parseResult.hasTag("dont")`

**Comment by TheLimeGlass** [comment]
Shouldn't it be included in the test event? What if we want to make a test run for a Paper exclusive syntax? This will just error each time.

**Comment by Pikachu920** [comment]
> Shouldn't it be included in the test event? What if we want to make a test run for a Paper exclusive syntax? This will just error each time. Or have another structure or section to parse only if the method exists?

it should work in the test event - Pickle said that the test event supports using any condition (this makes sense because `CondMinecraftVersion` is in the same package). The reason I have a test for it is just to make sure it's working as intended. It's not meant to demo how you would actually use it in a test event (but I can see how that would be confusing lol).

**Review by AyhamAl-Ali — CHANGES_REQUESTED** [review]
Amazing PR ⚡

File: src/main/java/ch/njol/skript/tests/runner/CondMethodExists.java (hunk 1 above)
Comment (AyhamAl-Ali) [reply src/main/java/ch/njol/skript/tests/runner/CondMethodExists.java:48]: the `'t` can probably be after the grouping as well

**Comment by TheLimeGlass** [comment]
You can place regex in the actual Skript syntax pattern `<your regex here>` instead of %string% and defining the field pattern.

**Comment by Pikachu920** [comment]
> You can place regex in the actual Skript syntax pattern `<your regex here>` instead of %string% and defining the field pattern.

hmmm... I think I would prefer to keep it this way as otherwise I'll have to do manual parsing for multiple strings. what do you think?

**Comment by Pikachu920** [comment]
Needs support for primitive params

**Comment by TheLimeGlass** [comment]
> > You can place regex in the actual Skript syntax pattern `<your regex here>` instead of %string% and defining the field pattern.
> 
> hmmm... I think I would prefer to keep it this way as otherwise I'll have to do manual parsing for multiple strings. what do you think?

It shouldn't support multiple strings. You should make a separate test structure if that's the case. Typically if one method doesn't exist that insists that the functionality Skript is testing for is not present.

**Comment by Pikachu920** [comment]
> > > You can place regex in the actual Skript syntax pattern `<your regex here>` instead of %string% and defining the field pattern.
> > 
> > 
> > hmmm... I think I would prefer to keep it this way as otherwise I'll have to do manual parsing for multiple strings. what do you think?
> 
> It shouldn't support multiple strings. You should make a separate test structure if that's the case. Typically if one method doesn't exist that insists that the functionality Skript is testing for is not present.

I agree that typically it's only one, but could you explain why it shouldn't support multiple strings please?

**Comment by TheLimeGlass** [comment]
> 
> I agree that typically it's only one, but could you explain why it shouldn't support multiple strings please?

I just did. A test shouldn't be taking two possible missing methods, it should be defined in a separate test structure.

File: src/main/java/ch/njol/skript/tests/runner/CondMethodExists.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+	static {
+		Skript.registerCondition(CondMethodExists.class, "method[s] %strings% [dont:do(esn't|n't)] exist[s]");
+	}
+
+	@SuppressWarnings("NotNullFieldNotInitialized")
+	private Expression<String> signatures;
+
+	@SuppressWarnings("unchecked")
+	@Override
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/tests/runner/CondMethodExists.java:56]:
```suggestion
	@Override
	@SuppressWarnings("unchecked")
```

**Review by AyhamAl-Ali — APPROVED** [review]
Approved so far. Will do another review after primitive params and arrays are supported ⚡

**Review by kiip1 — COMMENTED** [review]
Looks good, noticed one small thing.

**Review by TheLimeGlass — APPROVED** [review]

**Pull request merged**